src/features/virtualDocumentTracker.ts

```diff
@@ -0,0 +1,91 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import {workspace, TextDocument} from 'vscode';
+import {OmniSharpServer} from '../omnisharp/server';
+import * as serverUtils from '../omnisharp/utils';
+import { FileChangeType } from '../omnisharp/protocol';
+import { IDisposable } from '../Disposable';
+import CompositeDisposable from '../CompositeDisposable';
+
+function trackCurrentVirtualDocuments(server: OmniSharpServer) {
+    let registration = server.onProjectAdded(() => {
+        for (let i = 0; i < workspace.textDocuments.length; i++) {
+            let document = workspace.textDocuments[i];
+    
+            if (!shouldIgnoreDocument(document, server)) {
+                openVirtualDocument(document, server);
+            }
+        }
+
+        registration.dispose();
+    });
+}
+
+function trackFutureVirtualDocuments(server: OmniSharpServer): IDisposable {
+    let onTextDocumentOpen = workspace.onDidOpenTextDocument(document => {
+        if (shouldIgnoreDocument(document, server)) {
+            return;
+        }
+
+        openVirtualDocument(document, server);
+    });
+
+    let onTextDocumentClose = workspace.onDidCloseTextDocument(document => {
+        if (shouldIgnoreDocument(document, server)) {
+            return;
+        }
+
+        closeVirtualDocument(document, server);
+    });
+    
+    // We already track text document changes for virtual documents in our change forwarder.
+    return new CompositeDisposable(
+        onTextDocumentOpen,
+        onTextDocumentClose);
+}
+
+function shouldIgnoreDocument(document: TextDocument, server: OmniSharpServer): boolean {
+    if (document.uri.scheme === 'file' || document.languageId !== 'csharp') {
+        // We're only interested in non-physical CSharp documents.
+        return true;
+    }
+
+    if (!server.isRunning()) {
+        return true;
+    }
+
+    return false;
+}
+
+function openVirtualDocument(document: TextDocument, server: OmniSharpServer) {
+    let req = { FileName: document.uri.path, changeType: FileChangeType.Create };
+    serverUtils.filesChanged(server, [req])
+        .catch(err => {
+            console.warn(`[o] failed to forward virtual document change event for ${document.uri.path}`, err);
```

Comment:
@akshita31 Can you weigh in on the best way to do this logging?

Reply:
Yes, we should use the C# channel to do the logging here.
@NTaylorMullen The way logging works in the extension , is that we post a  [loggingEvent](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/omnisharp/loggingEvents.ts) to an [https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/EventStream.ts](event stream), to which many observers have susbsribed. When we post the event to the event stream, all the observers that have subscribed to the stream, perform the necessary action. For this case, we will have to add the appropriate logging event and make the [C# logger observer](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/observers/CsharpLoggerObserver.ts) listen to it. I am not sure if the [C# channel observer ](https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/observers/CsharpChannelObserver.ts) , should care of these events and show the channel. I will leave that up to you.

Reply:
@akshita31 How does this component get access to the eventStream?

Reply:
We can pass it down from the extension.ts - https://github.com/OmniSharp/omnisharp-vscode/pull/2436/files#diff-f49a39a20dc9807208c37823c89847a6R84